kernel32.cpp

```diff
@@ -577,11 +695,205 @@ namespace kernel32 {
+		DEBUG_LOG("WideCharToMultiByte(codePage=%u, flags=%x, wcs=%p, wideChar=%d, mbs=%p, multiByte=%d, defaultChar=%p, usedDefaultChar=%p)\n", codePage, dwFlags, lpWideCharStr, cchWideChar, lpMultiByteStr, cbMultiByte, lpDefaultChar, lpUsedDefaultChar);
+
+		if (cchWideChar == -1) {
+			// determine how long the string actually is
+			cchWideChar = 0;
+			while (lpWideCharStr[cchWideChar] != 0)
+				++cchWideChar;
+		}
+
+		if (cbMultiByte == 0) {
+			return cchWideChar + 1;
+		}
+		for (int i = 0; i < cchWideChar; i++) {
+			lpMultiByteStr[i] = lpWideCharStr[i];
+		}
+		lpMultiByteStr[cchWideChar] = 0;
+		DEBUG_LOG("Converted string: [%s]\n", lpMultiByteStr);
+
+		return cbMultiByte;
+	}
+
+	unsigned int WIN_FUNC FreeEnvironmentStringsW(void *penv) {
+		DEBUG_LOG("FreeEnvironmentStringsW: %p\n", penv);
+		free(penv);
+		return 1;
+	}
+
+	unsigned int WIN_FUNC IsProcessorFeaturePresent(unsigned int processorFeature) {
+		DEBUG_LOG("IsProcessorFeaturePresent: %u\n", processorFeature);
+
+		if (processorFeature == 0) // PF_FLOATING_POINT_PRECISION_ERRATA
+			return 1;
+
+		// sure.. we have that feature...
+		return 1;
+	}
+
+	void *WIN_FUNC GetProcAddress(void *hModule, char *lpProcName) {
+		DEBUG_LOG("GetProcAddress: %s from %p\n", lpProcName, hModule);
+
+		if ((unsigned int)hModule == 1) {
+			if (strcmp(lpProcName, "IsProcessorFeaturePresent") == 0) return (void *) IsProcessorFeaturePresent;
+		}
+
+		return NULL;
+	}
+
+	void *WIN_FUNC HeapAlloc(void *hHeap, unsigned int dwFlags, size_t dwBytes) {
+		DEBUG_LOG("HeapAlloc(heap=%p, flags=%x, bytes=%u)\n", hHeap, dwFlags, dwBytes);
+
+		void *mem = malloc(dwBytes);
+		if (mem && (dwFlags & 8))
+			memset(mem, 0, dwBytes);
+
+		DEBUG_LOG("HeapAlloc returning %p\n", mem);
+		return mem;
+	}
+
+	unsigned int WIN_FUNC HeapFree(void *hHeap, unsigned int dwFlags, void *lpMem) {
+		DEBUG_LOG("HeapFree(heap=%p, flags=%x, mem=%p)\n", hHeap, dwFlags, lpMem);
+		free(lpMem);
+		return 1;
+	}
+
+	unsigned int WIN_FUNC FormatMessageA(unsigned int dwFlags, void *lpSource, unsigned int dwMessageId,
+										 unsigned int dwLanguageId, char *lpBuffer, unsigned int nSize, char *arg0, ...) {
+
+		DEBUG_LOG("FormatMessageA: flags: %u, message id: %u\n", dwFlags, dwMessageId);
+
+		// TODO:
+		// va_list arguments;
+		// va_start(arguments, arg0);
+		// va_end(arguments);
+
+		if (dwFlags & 0x00000100) {
+			// FORMAT_MESSAGE_ALLOCATE_BUFFER
+		} else if (dwFlags & 0x00002000) {
+			// FORMAT_MESSAGE_ARGUMENT_ARRAY
+		} else if (dwFlags & 0x00000800) {
+			// FORMAT_MESSAGE_FROM_HMODULE
+		} else if (dwFlags & 0x00000400) {
+			// FORMAT_MESSAGE_FROM_STRING
+		} else if (dwFlags & 0x00001000) {
+			// FORMAT_MESSAGE_FROM_SYSTEM
+			std::string message = std::system_category().message(dwMessageId);
+			size_t length = message.length();
+			strcpy(lpBuffer, message.c_str());
+			return length;
+		} else if (dwFlags & 0x00000200) {
+			// FORMAT_MESSAGE_IGNORE_INSERTS
+		} else {
+			// unhandled?
+		}
```

Comment:
I think it's worth assert(0)ing in the unhandled cases, or at least null terminating lpBuffer